lib/moonshot/command_line.rb

```diff
@@ -0,0 +1,104 @@
+    def self.run!
+      # Find the Moonfile in this project.
+      orig_dir = Dir.pwd
+
+      loop do
+        break if File.exist?('Moonfile')
+
+        if Dir.pwd == '/'
+          warn 'Could not find Moonfile!'
+          exit 1
+        else
+          Dir.chdir('..')
+        end
+      end
+
+      moonfile_dir = Dir.pwd
+      Dir.chdir(orig_dir)
+
+      # Load any plugins and CLI extensions relative to the Moonfile
+      if File.directory?(File.join(moonfile_dir, 'moonshot'))
+        load_cli_extensions(moonfile_dir)
+        load_plugins(moonfile_dir)
+      end
+
+      Object.include(Moonshot::ArtifactRepository)
+      Object.include(Moonshot::BuildMechanism)
+      Object.include(Moonshot::DeploymentMechanism)
+      load(File.join(moonfile_dir, 'Moonfile'))
+
+      # Inject new tasks into CommandLine, as defined in the Moonfile.
+      # Moonshot.config.tasks.each do |t|
+      #   CommandLine.class_eval do
+      #     desc t.name, t.desc
+      #     define_method(t.name.to_s.gsub('-', '_'), &t.block)
+      #   end
+      # end
+
+      # Moonshot.config.ssh_tasks.each do |t|
+      #   CommandLine.class_eval do
+      #     option :user, default: ENV['MOONSHOT_SSH_USER'] || ENV['USER'], type: :string,
+      #                   aliases: '-l',
+      #                   desc: 'Specifies the user to log in as on the remote machine.'
+      #     option :identity_file, default: ENV['MOONSHOT_SSH_KEY_FILE'], type: :string,
+      #                            aliases: '-i',
+      #                            desc: 'Selects a file from which the identity (private key) for '\
+      #                                  'public key authentication is read.'
+      #     option :instance, default: nil, type: :string, aliases: '-s',
+      #                      desc: 'Connect to specified instance ID instead of the first one.'
+      #     option :command, default: nil, type: :string, aliases: '-c',
+      #                      desc: 'Execute the specified command instead of opening a shell.'
+      #     option :auto_scaling_group, default: nil, type: :string, aliases: '-g',
+      #                                 desc: 'The logical ID of the auto scaling group to SSH into. '\
+      #                                       'Mandatory if the stack has multiple ASGs, '\
+      #                                       'ignored if you have only one.'
+      #     desc t.name, t.desc
+      #     define_method(t.name.to_s.gsub('-', '_'), &t.block)
+      #   end
+      # end
+
+      # Include all Moonshot::Command and Moonshot::SSHCommand
+      # derived classes as subcommands, with the description of their
+      # default task.
+      registered_commands.each do |klass|
+        default_task = klass.default_task
+        description = klass.tasks.fetch(default_task).description rescue ''
+        command_name = klass.to_s.split('::').last.scan(/[A-Z][a-z]+/).map(&:downcase).join('-')
+
+        desc command_name, description
```

Comment:
Well, you can race me to replacing this with OptionParser. Whomever gets there first wins. :-)